### PR TITLE
[SPARK-41398][SQL] Relax constraints on Storage-Partitioned Join when partition keys after runtime filtering do not match

### PR DESCRIPTION
**Pull request by sunchao**
### What changes were proposed in this pull request?

This PR relaxes the current constraint of Storage-Partitioned Join which requires that the partition keys after runtime filtering to be exact the same as the partition keys before the filtering. 

### Why are the changes needed?

At the moment, Spark requires that when Storage-Partitioned Join is used together with runtime filtering, the partition keys before and after the filtering shall exact match. If not, a `SparkException` is thrown.

However, this is not strictly necessary in the case where the partition keys after the filtering is a subset of the original keys. In this scenario, we can use empty partitions for those missing keys in the latter.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Modified an existing test case to match the change.

### CODE REVIEW

**Comment by sunchao** [comment]
cc @cloud-fan @dongjoon-hyun @viirya @aokolnychyi @huaxingao

**Review by dongjoon-hyun — COMMENTED** [review]
Thank you for making a PR, @sunchao .

**Review by dongjoon-hyun — APPROVED** [review]
From my side, it looks good to me.

**Comment by dongjoon-hyun** [comment]
Also, cc @wangyum , @MaxGekk , @yaooqinn , @c21

**Comment by dongjoon-hyun** [comment]
Merged to master for Apache Spark 3.4.0.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala [hunk 1]
```diff
@@ -81,18 +81,21 @@ case class BatchScanExec(
 
           val newRows = new InternalRowSet(p.expressions.map(_.dataType))
           newRows ++= newPartitions.map(_.asInstanceOf[HasPartitionKey].partitionKey())
-          val oldRows = p.partitionValuesOpt.get
 
-          if (oldRows.size != newRows.size) {
-            throw new SparkException("Data source must have preserved the original partitioning " +
-                "during runtime filtering: the number of unique partition values obtained " +
-                s"through HasPartitionKey changed: before ${oldRows.size}, after ${newRows.size}")
+          val oldRows = p.partitionValuesOpt.get.toSet
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala:85]: shall we use `InternalRowSet`?

Comment (sunchao) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala:85]: Oops you're right! I forgot about `InternalRowSet`. Let me create a follow-up.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala [hunk 2]
```diff
@@ -114,8 +117,21 @@ case class BatchScanExec(
       // return an empty RDD with 1 partition if dynamic filtering removed the only split
       sparkContext.parallelize(Array.empty[InternalRow], 1)
     } else {
+      var finalPartitions = filteredPartitions
+
+      outputPartitioning match {
+        case p: KeyGroupedPartitioning =>
+          val partitionMapping = finalPartitions.map(s =>
+            s.head.asInstanceOf[HasPartitionKey].partitionKey() -> s).toMap
+          finalPartitions = p.partitionValuesOpt.get.map { partKey =>
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala:126]: shall we move this logic to `lazy val filteredPartitions`?

Comment (sunchao) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala:126]: Hmm any obvious advantage of this? It looks the same to me 😅

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala:126]: It's better to centralize the related code.

Comment (sunchao) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala:126]: I was hoping to re-use the logic code in #38950 too, which is not related to runtime filtering. If moving this close to the `if (oldRows.size < newRows.size)` check then I may have to duplicate it somewhere else?

**Pull request closed without merging**